### ISSUE TITLE
Endlessly resizes the window on Classilla on OS 9

### ISSUE THREAD

**Issue by AbsolutelyLudicrous** [ttalvitie/browservice#33]
Hello there!

I'm connecting to browservice from Classilla on MacOS 9. When Classilla loads the page, it places a horizontal scroll bar at the bottom of the page, which shrinks Browservice's viewport. Browservice detects this, and resizes its viewport to be smaller. Classilla then removes the scrollbar since the viewport is now small enough to fit in the window, which causes Browservice to try and fill up the extra bit of space, which causes the scrollbar to reappear, which causes Browservice to shrink, which causes the scrollbar to disappear, on and on forever. This may be the same bug as #32, though where as #32 just shows a blank page and the URL keeps changing, here we get a seizure-inducing loop of the selected homepage rerendering itself.

Other than that, this works great on MacOS 9.2.2/Classilla 9.3.4b

**Comment by brunocastello** [comment]
I am not getting that behavior. I get just a white page with nothing.

**Comment by brunocastello** [comment]
I had to enable Javascript globally on Classilla to be able to reproduce this behavior. Before I did not even get a browservice address bar. I think that we need to find a way to disable the scrollbars on Classilla to fix this. Classilla does not have a userChrome.css nor any way to customize CSS per page.

**Comment by ttalvitie** [comment]
I remember coming across this resizing issue on various browsers; probably at least on classic Mozilla on Debian Sarge. As you said, it is caused by both Browservice and the browser reacting to image size and scroll bar visibility changes in a feedback loop. Possible solutions include:

- Forcing the hiding of scrollbars. Browservice tries to do this using CSS `body { overflow: hidden; }`, which is not supported on all browsers.
- Detecting the browser window size such that no scrollbars are created. This is highly browser-dependent.

On some browsers (IE4 and IE5 on Win3.x), I was able to hack around this problem by reducing the detected width and height by 1 and 2 pixels, respectively (this way, the image does not overflow the page; normally the image overflows for signaling purposes). You could try whether this works in your case as well. To do this, add the line `width -= 20; height -= 20;` before line 198 in [viceplugins/retrojsvice/html/main.html](https://github.com/ttalvitie/browservice/blob/v0.9.2.0/viceplugins/retrojsvice/html/main.html#L198). If this works, try smaller numbers instead of `20`, for example `width -= 1; height -= 2;`. Note that you need to update the Browservice executable after each change by running `make`, because the HTML templates are built into the executable.

**Comment by brunocastello** [comment]
I’ve just tried the fix I made following your instructions.

Apparently it was already set up for width -= 1; and height -= 2;

So I tried 20/20, no change at all. I decided to revert it back to original values.

However we do have a userContent.css and userChrome.css, I’ve spent hours trying to get the width right. However it seems to be an impossible mission, because no matter how much I get the maths right, there is always one pixel left for the width. Weird.

I’ll try other CSS methods. Meanwhile... do you have other ideas?

**Comment by ttalvitie** [comment]
I'm sorry, my instructions were a bit unclear. By "before line 198" I meant "between lines 197 and 198", which means that the added lines should NOT be inside the `if(bodyOverflowHiddenNotSupported) { ... }` block (`bodyOverflowHiddenNotSupported` is only `true` for IE4 and IE5 on Win3.x, and thus if you make changes inside the block it does not actually affect anything). Make sure to reduce the width and height after the ending `}` for the `if` block and before `width = Math.max(width, 1);`.

The reason that the width is larger than the browser window is that the width and height are used for signaling (from the server to the client; things like mouse cursors and requests to load invisible iframes for example for file downloads). This is currently the only way to pass signals to the client JavaScript (as we do not require AJAX).

**Comment by ttalvitie** [comment]
So it should look like this (in retrospect, maybe I should have just pushed this to a branch):
```
...
    if(bodyOverflowHiddenNotSupported) {
        width -= 1;
        height -= 2;
    }

    width -= 20;
    height -= 20;

    width = Math.max(width, 1);
...
```

**Comment by brunocastello** [comment]
I managed to work around the issue using the userContent.css, by giving the body a position absolute and zeros for all four corners along with a minimal CSS to get it working.

@ttalvitie I will try your solution in some minutes and see if it helps so you can check if this should be pushed or not. 

Meanwhile I have also made the change for IE 5 on MacOS 9, but I havent tried it yet because I still havent installed IE 5.

**Comment by brunocastello** [comment]
20/20 works, but there is a big space left on both right and bottom sides. Gotta try and reduce it as much as I can until its fine again. I used if else instead of changing the values after the block. Because when I used Browservice with my Windows vintage systems it was fine as is. Now I will try again and reduce the values to see how far I can go.

**Comment by brunocastello** [comment]
New report:

I tried width -= 2; and height -= 2; and there was a small 1px border on the right. Close to the solution, yet so far from ideal.

Using width -= 1; returns to the original problem. Kinda strange.

Think I will stick to the userContent.css solution for the time being.

**Comment by ttalvitie** [comment]
Reducing width and height has its side effects, because it adds small white areas in the right and bottom of the browser windows. So we should only do it on browsers that actually need it. Thus I think a8bf95015091474b9542abc7c3d9e24a84f6dc17 should fix it. Could you confirm that the fix works? You can do this by replacing `viceplugins/retrojsvice/html/main.html` with the [updated version](https://github.com/ttalvitie/browservice/blob/overflow-classilla-fix/viceplugins/retrojsvice/html/main.html). Remember to recompile using `make`.

**Comment by brunocastello** [comment]
My report as of now is:

Setting them to width -= 1; and height -= 2; as you previously mentioned, works for both IE 5.1.7 and Classilla 9.3.4b perfectly under Mac OS 9.2.2 (QEMU VM). So good that I haven’t changed again for more tests and set IE 5.1.7 to not show any toolbars, apart of back, forward, home and reload (and Browservice’s toolbar). I don’t know if this change affected my other Windows VMs, but apparently Win 9x IE6 is still working OK.

I can test a8bf950 later, but Classilla uses a nokia user agent, because Cameron Kaiser thinks it works best for old macs to deliver mobile web pages. Do you want me to show which other user agents are selectable from Classila’s preference panes?

Classilla can use Browservice too (I have to enable global JS for it to work), but I currently had set Classilla to use Cameron Kaiser’s Crypto Ancienne to give it TLS 1.2 support as a backup browser, for some stuff like https downloads when I cannot find what I want from Macintosh Garden, for example. (But this rarely happens, Mac Garden always has everything). Although I think that old browsers with TLS 1.2 support are still “useless” when their rendering engines are now useless for Web 2.0 (Netscape 9 and Firefox > 35 ESR are still quite decent though). This is where Browservice becomes a great addition.

I haven’t tried yet the updated version. Does this new version assume that it will always be these values for both browsers on Mac OS 9? Because these are the only browsers that needed this change so far. Current version works great for for my Win 9x browsers (IE6 and Netscape 9). I haven’t tested yet with Windows For Workgroups 3.11 VM.

**Comment by ttalvitie** [comment]
The fix a8bf95015091474b9542abc7c3d9e24a84f6dc17 tries to detect only Classilla based on the user agent string, but based on what you say it probably won't work, as Classilla spoofs its user agent. So if there is no other portable way to detect Classilla, the only option that comes to mind would be to have command line arguments for enabling `width -= 1; height -= 2;`.

IE5 is probably unrelated to this, as it worked without any fixes (apart from the redirect loop thing), right?

**Comment by brunocastello** [comment]
> The fix [a8bf950](https://github.com/ttalvitie/browservice/commit/a8bf95015091474b9542abc7c3d9e24a84f6dc17) tries to detect only Classilla based on the user agent string, but based on what you say it probably won't work, as Classilla spoofs its user agent. So if there is no other portable way to detect Classilla, the only option that comes to mind would be to have command line arguments for enabling `width -= 1; height -= 2;`.
> 
> IE5 is probably unrelated to this, as it worked without any fixes (apart from the redirect loop thing), right?

1) Yup. Classilla initially spoofs a mobile nokia based user agent string by default, out of the box. It can be changed to other user agent strings like Firefox 35 ESR as far as I can remember now. You can, though, treat it just like another mozilla browser, (Classila is based on Mozilla 1.3x code) and tell users to spoof one of the available user agent strings on Classilla if they want to use it. I think it might be easier for them.

Netscape 7.0.2 though, the last version to work on OS 9, did not work. Apparently, it's a javascript issue. I tried the same tweaks I did for Classilla, but no luck.

2) Yup, apart of the redirect loop, IE5 works well. I still haven't tried your modifications but I plan to do so this weekend, it's when I have more time to do these things. Maybe friday night I can have a head start on it.

**Comment by brunocastello** [comment]
Hi @ttalvitie, below are some screenshots expliciting the default and available user agents for Classilla 9.3.4b (last version, since Cameron Kaiser is not going to develop anymore and everyone on macOS 9 should be using this for now)

![IMG_0449](https://user-images.githubusercontent.com/23453202/116722157-72ca4c80-a9b4-11eb-9c54-b0d60d4d7f20.png)
Default Classilla user agent (Nokia N90 user agent)

![IMG_0448](https://user-images.githubusercontent.com/23453202/116722646-0b60cc80-a9b5-11eb-8b55-afe926d415d5.png)
Current available user agents for Classilla in preference panes

Surely by just using MSIE6 (Windows) user agent we could overcome this? But users would have to come to this screen and switch back to Nokia N90 whenever they wanted a native experience without Browservice. Not an ideal user experience. We might want to try and use the default Classilla user agent to detect this. Let me know if we can use this and how to change the #a8bf950 for this user agent.

**Comment by ttalvitie** [comment]
I don't understand why using MSIE6 (Windows) user agent would work; on Classilla we need `width -= 1; height -= 2;`, on IE6 on Window we don't.

However, from the first picture it appears that the default Classilla user agent string actually contains the word "Classilla" at the end, so a8bf95015091474b9542abc7c3d9e24a84f6dc17 should work correctly; please confirm by testing.

**Comment by brunocastello** [comment]
hi @ttalvitie , 

I can confirm that [a8bf950](https://github.com/ttalvitie/browservice/commit/a8bf95015091474b9542abc7c3d9e24a84f6dc17) fixes the issue for Classilla. It recognized the user agent used by Classilla. I recompiled and reloaded it. However, several unrelated stability issues surfaced. I used the branch disable-mac-ie-native-navigation for this test.

Classilla started to crash a few times, notably when I always tried the same site - https://www.macrumors.com/ , and Browservice also crashed a few times in the proccess, I had to restart it once. But the fix for Classilla works, I will apply the fix to the master branch to continue testing. I believe these new issues are related to the fix you made for IE5 on MacOS9.

The fix for IE5 on MacOS9 though, did not work, just worsened it. Browservice now simply refuses to work on IE5 and keeps the toolbar buttons blinking non-stop (back, forward, stop buttons). I'll be reverting to the master branch and apply the fix that worked for Classilla here, so I can continue using it for both browsers. I think you can close this issue and we can continue on #32 talking about fixing it for IE5.1.7 on OS9.

**Comment by ttalvitie** [comment]
The IE5 fix should not affect this, as it is not included in the commit a8bf95015091474b9542abc7c3d9e24a84f6dc17 (that is, the tip of the branch overflow-classilla-fix). From now on, I recommend that you test these fix proposal branches without manual modifications, by checking out the correct branch in a working copy. I'm also very interested in what happens in these crashes; to get better output, you should do a debug build. To do this, run the following:

```
git clone https://github.com/ttalvitie/browservice browservice-testing
cd browservice-testing

git checkout overflow-classilla-fix

# Install CEF only after checking out overflow-classilla-fix, as the current master (0.9.2.1) has newer CEF than 0.9.2.0 on which overflow-classilla-fix is based
./download_cef.sh
./setup_cef.sh

make -j5 debug
sudo chown root:root debug/bin/chrome-sandbox && sudo chmod 4755 debug/bin/chrome-sandbox

# Run Browservice; add command line options if necessary
debug/bin/browservice
```

**Comment by ttalvitie** [comment]
Closing due to lack of activity; please create a new issue if you still have this problem.

**Comment by asdsadadad** [comment]
Was this fix ever added to the master branch? I'm still getting this error on v0.9.6.3 on Windows.
